google/cloud/bigtable/batcher.py

```diff
@@ -125,22 +298,120 @@ def mutate_rows(self, rows):
                    match the number of rows that were retried
                  * :exc:`.batcher.MaxMutationsError` if any row exceeds max
                    mutations count.
+                 * :exc:`.batcherMutationsBatchError` if there's any error in the
+                   mutations.
         """
         for row in rows:
             self.mutate(row)
 
     def flush(self):
-        """Sends the current. batch to Cloud Bigtable.
+        """Sends the current batch to Cloud Bigtable synchronously.
         For example:
 
         .. literalinclude:: snippets.py
             :start-after: [START bigtable_api_batcher_flush]
             :end-before: [END bigtable_api_batcher_flush]
             :dedent: 4
 
+        raises:
+            * :exc:`.batcherMutationsBatchError` if there's any error in the
+                mutations.
+        """
+        rows_to_flush = []
+        while not self._rows.empty():
+            rows_to_flush.append(self._rows.get())
+        response = self.flush_rows(rows_to_flush)
+        return response
+
+    def flush_async(self):
+        """Sends the current batch to Cloud Bigtable asynchronously.
+
+        raises:
+            * :exc:`.batcherMutationsBatchError` if there's any error in the
+                mutations.
+        """
+
+        rows_to_flush = []
+        mutations_count = 0
+        mutations_size = 0
+        rows_count = 0
+        batch_info = BatchInfo()
+
+        while not self._rows.empty():
+            row = self._rows.get()
+            mutations_count += len(row._get_mutations())
+            mutations_size += row.get_mutations_size()
+            rows_count += 1
+            rows_to_flush.append(row)
+            batch_info.mutations_count = mutations_count
+            batch_info.rows_count = rows_count
+            batch_info.mutations_size = mutations_size
+
+            if (
+                rows_count >= self.flush_count
+                or mutations_size >= self.max_row_bytes
+                or mutations_count >= self.flow_control.max_mutations
```

Comment:
I dont think we need to check flow control limits here since you're calling `self.flow_control.control_flow(batch_info)` on line 357.

Reply:
We still need to check it here. At this point the flow control has not been activated yet. We're checking if we will need to activate the flow control after this condition is met. Without this check it will not actually activate the flow control.